public/recorder.js

```diff
@@ -41,9 +56,186 @@ coverImg.appendChild(avatar);
 coverTextInp.value = randomEmoji;
 
 
+function resetAudioGraph() {
+  if (playbackNode) {
+    playbackNode.disconnect();
+    playbackNode.removeEventListener("ended", onEndedHandler);
+    playbackNode = null;
+  }
+  if (refNode) {
+    refNode.disconnect();
+    refNode.removeEventListener("ended", onEndedHandler);
+    refNode = null;
+  }
+}
+
+function onEndedHandler() {
+  audioContext.suspend();
+  resetAudioGraph();
+  stopKaraoke();
+
+  if (recorder) {
+    recorder.stop();
+    stopLabel.textContent = "Stop";
+  }
+
+  recordBtn.disabled = false;
+  playBtn.disabled = false;
+  pauseBtn.disabled = true;
+  stopBtn.disabled = true;
+}
+
+
+async function prepareReferencePlayback() {
+  startTime = audioContext.currentTime;
+  playbackNode = audioContext.createBufferSource();
+  playbackNode.buffer = refDecodedAudio;
+  playbackNode.start(startTime);
+  playbackNode.connect(audioContext.destination);
+  playbackNode.addEventListener("ended", onEndedHandler);
+}
+
+async function prepareCountdownAndReferencePlayback() {
+  startTime = audioContext.currentTime;
+  playbackNode = audioContext.createGain();
+
+  const toneNode1 = audioContext.createBufferSource();
+  toneNode1.buffer = toneDecodedAudio;
+  toneNode1.start(startTime);
+  toneNode1.stop(startTime + 1);
+  toneNode1.connect(playbackNode);
+
+  const toneNode2 = audioContext.createBufferSource();
+  toneNode2.buffer = toneDecodedAudio;
+  toneNode2.start(startTime + 1);
+  toneNode2.stop(startTime + 2);
+  toneNode2.connect(playbackNode);
+
+  const toneNode3 = audioContext.createBufferSource();
+  toneNode3.buffer = toneDecodedAudio;
+  toneNode3.start(startTime + 2);
+  toneNode3.stop(startTime + 3);
+  toneNode3.connect(playbackNode);
+
+  refNode = audioContext.createBufferSource();
+  refNode.buffer = refDecodedAudio;
+  refNode.start(startTime + 3);
```

Comment:
The new behavior seems to be playing the tone 3 times, instead of the 2 seconds intro tone followed by a second of silence - I'm not sure this an improvement. Is this by design? what is the intent behind that change?

Reply:
I added this as I realized that a visual 3.. 2.. 1.. clue was not enough for me to get the tempo and starting point right. I often missed the beginning of the song as a result. Ideally, there would be the first note followed by 3... 4.. and the music would start. YMMV :) I can drop that if you prefer.

Reply:
I would prefer if we moved this to a separate PR at least to discuss the pros & cons